File: apps/macos/Sources/NaluVoiceStudio/EpisodeAudioReview.swift
import Foundation
import Observation

struct EpisodeAudioReviewRecovery: Decodable, Sendable {
    let current_take_id: String
    let current_take_sha256: String
    let latest_review: EpisodeAudioReview?
    let applies_to_current_take: Bool
    let take_approved: Bool
}

struct EpisodeAudioReviewDraft: Encodable, Sendable {
    enum Decision: String, Codable, Sendable { case accept, reject }
    let expected_take_sha256: String
    let expected_review_id: String?
    let decision: Decision
    let reviewed_by: String
    let confirmation: String
}

struct EpisodeAudioReview: Decodable, Sendable {
    let id: String
    let run_id: String
    let event_type: String
    let payload: Payload

    struct Payload: Decodable, Sendable {
        let take_id: String
        let take_sha256: String
        let sound_plan_id: String
        let sound_plan_sha256: String
        let edit_review_id: String
        let edit_sha256: String
        let shot_index: Int
        let asset_id: String
        let asset_sha256: String
        let source_in_seconds: Double
        let duration_seconds: Double
        let decision: EpisodeAudioReviewDraft.Decision
        let reviewed_by: String
        let confirmation: String
        let review_sha256: String
        let take_approved: Bool
        let listening_evidence: String
        let speech_alignment_verified: Bool
        let final_mix_approved: Bool
        let captions_approved: Bool
        let master_accepted: Bool
        let generation_performed: Bool
    }
}

@MainActor @Observable final class EpisodeAudioReviewModel {
    let sound: EpisodeSoundPlan
    let take: EpisodeAudioTake
    private let runtime: RuntimeClient
    private(set) var busy = false
    private(set) var loaded = false
    private(set) var latest: EpisodeAudioReviewRecovery?
    private(set) var pending: EpisodeAudioReviewDraft?
    private(set) var uncertain = false
    private(set) var transcript: EpisodeRecordingTranscript?
    private(set) var transcriptReceipt: RecordingTranscriptRecord?
    private(set) var notice = "请先试听这段录音，再确认是否采用。还需要字幕和成片检查。"

    init(sound: EpisodeSoundPlan, take: EpisodeAudioTake, runtime: RuntimeClient = RuntimeClient()) {
        self.sound = sound; self.take = take; self.runtime = runtime
    }

    var readback: String {
        let window = "从第 \(take.payload.source_in_seconds.formatted()) 秒开始的约 \(take.payload.duration_seconds.formatted()) 秒录音"
        return pending?.decision == .accept
            ? "您确认已试听并采用\(window)。还需要字幕、混音和成片检查，不会自动发行。确认吗？"
            : "您要退回\(window)。原录音保留，不会自动生成或付费重做。确认吗？"
    }

    func load() async {
        guard !busy else { return }
        busy = true; loaded = false
        defer { busy = false }
        do {
            let recovered = try await runtime.recoverEpisodeAudioReview(sound: sound, take: take)
            guard !Task.isCancelled else { return }
            latest = recovered; loaded = true
            if !recovered.take_approved || transcript?.reviewID != recovered.latest_review?.id { transcript = nil }
            if !recovered.take_approved || transcriptReceipt?.payload.expected_review_id != recovered.latest_review?.id {
                transcriptReceipt = nil
            }
            if let pending, uncertain, recovered.applies_to_current_take,
               let review = recovered.latest_review,
               review.payload.decision == pending.decision,
               review.payload.reviewed_by == pending.reviewed_by,
               review.payload.confirmation == pending.confirmation {
                self.pending = nil; uncertain = false
                notice = "已找回刚才的确认结果，不需要重复提交。"
            } else if uncertain {
                notice = "刚才的确认结果尚未核实，请保留同一选择重试；原录音不变。"
            } else if recovered.take_approved {
                notice = "这段录音已采用；字幕、混音和成片检查仍未完成。"
            } else {
                notice = "当前录音尚未采用。请试听后确认；旧录音的确认不会自动沿用。"
            }
        } catch { notice = "暂时无法核对录音确认结果。请重试，原素材和待确认选择都保留。" }
    }

    func begin(_ decision: EpisodeAudioReviewDraft.Decision) -> Bool {
        guard loaded, !busy, pending == nil, let latest else { return false }
        pending = EpisodeAudioReviewDraft(expected_take_sha256: take.payload.take_sha256,
            expected_review_id: latest.latest_review?.id, decision: decision,
            reviewed_by: "local-user", confirmation: decision == .accept
                ? "我已试听并确认采用这一段录音。" : "这段录音需要调整，暂不采用。")
        return true
    }

    func cancelUnsubmitted() {
        guard !busy, !uncertain else { return }
        pending = nil
    }

    func confirm() async {
        guard !busy, let pending else { return }
        busy = true; uncertain = true
        defer { busy = false }
        do {
            let review = try await runtime.reviewEpisodeAudio(sound: sound, take: take, draft: pending)
            latest = EpisodeAudioReviewRecovery(current_take_id: take.id, current_take_sha256: take.payload.take_sha256,
                latest_review: review, applies_to_current_take: true, take_approved: review.payload.take_approved)
            transcript = nil
            transcriptReceipt = nil
            self.pending = nil; uncertain = false; loaded = true
            notice = review.payload.take_approved
                ? "录音采用已保存。接下来核对字幕和混音，不会自动发行。"
                : "已记录需要调整。原录音和剪辑都保留。"
        } catch { notice = "确认是否保存还未核实。请先核对保存结果，或重试同一确认。" }
    }

    func prepareTranscript() async {
        guard loaded, !busy, pending == nil, !uncertain, latest?.take_approved == true,
              transcript == nil || transcriptReceipt != nil,
              let reviewID = latest?.latest_review?.id else { return }
        busy = true; transcript = nil; transcriptReceipt = nil
        notice = "正在用本机识别已采用录音，准备带时间的字幕草稿；不会上传云端。"
        defer { busy = false }
        do {
            let audio = try await runtime.downloadAcceptedEpisodeAudio(sound: sound, take: take, expectedReviewID: reviewID)
            let draft = try await EpisodeRecordingTranscriber().transcribe(audio)
            // Recognition can take time: never adopt results after a changed
            // recording decision or revoked source authorization.
            let recovered = try await runtime.recoverEpisodeAudioReview(sound: sound, take: take)
            guard !Task.isCancelled, recovered.take_approved,
                  recovered.latest_review?.id == reviewID else { throw LibrarySnapshotRefreshError.contextChanged }
            latest = recovered; transcript = draft
            transcriptReceipt = try await runtime.saveRecordingTranscript(take: take, draft: draft)
            notice = "带时间的字幕草稿已保存，请核对实际说话内容。还没有确认字幕或成片。"
        } catch {
            notice = transcript == nil
                ? "本机字幕识别未完成，或录音确认已变化。原录音保留；核对记录后可重试，不会改用云端。"
                : "字幕草稿已识别，但保存结果还未核实。草稿保留，请重试保存，不必重新识别。"
        }
    }

    func savePendingTranscript() async {
        guard !busy, let transcript, pending == nil, !uncertain else { return }
        busy = true
        defer { busy = false }
        do {
            transcriptReceipt = try await runtime.saveRecordingTranscript(take: take, draft: transcript)
            notice = "字幕草稿已保存，接下来核对内容和时间。"
        } catch { notice = "字幕保存结果尚未核实，草稿保留，可以重试同一份草稿。" }
    }

    func recoverTranscript() async {
        guard loaded, !busy, pending == nil, !uncertain, latest?.take_approved == true,
              let reviewID = latest?.latest_review?.id else { return }
        busy = true
        defer { busy = false }
        do {
            let saved = try await runtime.recoverRecordingTranscript(take: take, reviewID: reviewID)
            guard !Task.isCancelled else { return }
            transcriptReceipt = saved
            notice = saved == nil ? "这一版录音还没有保存的字幕草稿，可以开始本机识别。"
                : "已恢复保存的字幕草稿；请继续核对内容和时间。"
        } catch { notice = "字幕草稿暂时无法恢复，原录音和已保存记录保留。" }
    }
}

File: services/runtime_api/nalu_runtime/episode_audio_review.py
"""Explicit listening decisions on exact recording windows, not final mix QA."""

import hashlib
import io
import sys
import time
import wave
from contextlib import nullcontext
from pathlib import Path
from typing import Literal
from urllib.parse import unquote, urlparse

from pydantic import BaseModel, ConfigDict, Field

from .episode_audio import EpisodeAudioService
from .models import RunEvent
from .postproduction_materializer import PostproductionMaterializationError, _audio_chunks
from .repository import ConflictError, encode, new_id, utc_now
from .video_preparation import digest


class EpisodeAudioReviewRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)
    expected_take_sha256: str = Field(pattern=r"^[a-f0-9]{64}$")
    expected_review_id: str | None = Field(default=None, min_length=1, max_length=160)
    decision: Literal["accept", "reject"]
    reviewed_by: str = Field(min_length=1, max_length=160)
    confirmation: str = Field(min_length=1, max_length=2000)


class EpisodeAudioReviewRecovery(BaseModel):
    current_take_id: str
    current_take_sha256: str
    latest_review: RunEvent | None
    applies_to_current_take: bool
    take_approved: bool


class EpisodeAudioReviewService:
    def __init__(self, repository, data_root):
        self.repository, self.data_root = repository, data_root

    def recover(self, run_id, take_id, expected_take_sha256, *, _db=None):
        repo = self.repository
        with (nullcontext(_db) if _db is not None else repo.db.connect()) as db:
            # Serialize the source/consent check and decision snapshot without
            # appending an event or replaying the user's confirmation.
            if _db is None:
                db.execute("BEGIN IMMEDIATE")
            take = repo.get_run_event(take_id)
            p = take.payload
            if (take.run_id != run_id or take.event_type != "episode_audio_take_attached"
                    or p.get("take_sha256") != expected_take_sha256
                    or digest({k: v for k, v in p.items() if k != "take_sha256"}) != expected_take_sha256):
                raise ConflictError("listening recovery requires the exact recording take")
            current = EpisodeAudioService(repo, self.data_root).recover(
                run_id, p["sound_plan_id"], p["expected_sound_plan_sha256"], _db=db)
            if not any(item.id == take_id for item in current):
                raise ConflictError("recording changed; recover the current take first")
            reviews = [event for event in repo.list_run_events(run_id)
                       if event.event_type == "episode_audio_take_reviewed"
                       and event.payload.get("sound_plan_id") == p["sound_plan_id"]
                       and event.payload.get("shot_index") == p["shot_index"]]
            latest = reviews[-1] if reviews else None
            applies = False
            if latest:
                r = latest.payload
                if r.get("review_sha256") != digest({k: v for k, v in r.items() if k != "review_sha256"}):
                    raise ConflictError("saved listening review integrity failed")
                reviewed_take = repo.get_run_event(r["take_id"])
                original = reviewed_take.payload
                if (reviewed_take.run_id != run_id or reviewed_take.event_type != "episode_audio_take_attached"
                        or r["take_sha256"] != original.get("take_sha256")
                        or original.get("take_sha256") != digest({k: v for k, v in original.items() if k != "take_sha256"})
                        or r["sound_plan_id"] != original.get("sound_plan_id")
                        or r["shot_index"] != original.get("shot_index")
                        or any(r.get(review_key) != original.get(take_key) for review_key, take_key in (
                            ("sound_plan_sha256", "expected_sound_plan_sha256"),
                            ("edit_review_id", "edit_review_id"), ("edit_sha256", "edit_sha256"),
                            ("asset_id", "asset_id"), ("asset_sha256", "expected_asset_sha256"),
                            ("source_in_seconds", "source_in_seconds"), ("duration_seconds", "duration_seconds")))
                        or r.get("listening_evidence") != "USER_ATTESTATION_NOT_PLAYBACK_TELEMETRY"
                        or r["decision"] not in {"accept", "reject"}
                        or r["take_approved"] != (r["decision"] == "accept")
                        or any(r.get(key) is not False for key in (
                            "speech_alignment_verified", "final_mix_approved", "captions_approved",
                            "master_accepted", "generation_performed"))):
                    raise ConflictError("saved listening review belongs to another recording")
                applies = r["take_id"] == take_id and r["take_sha256"] == expected_take_sha256
            # An older take's decision is retained only as the CAS predecessor,
            # never as approval of the replacement recording.
            return EpisodeAudioReviewRecovery(current_take_id=take_id, current_take_sha256=expected_take_sha256,
                latest_review=latest, applies_to_current_take=applies,
                take_approved=bool(applies and latest.payload.get("take_approved") is True))

    def accepted_audio(self, run_id, take_id, expected_take_sha256, expected_review_id):
        """Decode the exact accepted source window, without padding or cloning."""
        repo = self.repository
        with repo.db.connect() as db:
            db.execute("BEGIN IMMEDIATE")
            recovered = self.recover(run_id, take_id, expected_take_sha256, _db=db)
            if (not recovered.take_approved or recovered.latest_review is None
                    or recovered.latest_review.id != expected_review_id):
                raise ConflictError("confirm the current recording before preparing its audio")
            take = repo.get_run_event(take_id).payload
            asset = repo.get_asset(take["asset_id"])
            source = Path(unquote(urlparse(asset.local_uri).path))
            output = io.BytesIO()
            deadline = time.monotonic() + 30
            count = 0
            try:
                with wave.open(output, "wb") as target:
                    target.setnchannels(2)
                    target.setsampwidth(2)
                    target.setframerate(48000)
                    for samples in _audio_chunks(source, start_seconds=take["source_in_seconds"],
                            sample_count=take["decoded_sample_count"], require_full_duration=True,
                            should_cancel=lambda: time.monotonic() > deadline):
                        count += len(samples) // 2
                        if count > 300 * 48000:
                            raise ConflictError("accepted recording exceeds export duration")
                        if sys.byteorder != "little":
                            samples.byteswap()
                        target.writeframesraw(samples.tobytes())
            except PostproductionMaterializationError as exc:
                raise ConflictError("accepted recording could not be decoded") from exc
            # Recovery has already checked managed path and live consent; detect
            # a file edit during decode before returning any bytes to the caller.
            with source.open("rb") as current_source:
                source_bytes = current_source.read(100 * 1024 * 1024 + 1)
            if (count != take["decoded_sample_count"] or source.resolve() != source
                    or len(source_bytes) > 100 * 1024 * 1024
                    or hashlib.sha256(source_bytes).hexdigest() != take["expected_asset_sha256"]):
                raise ConflictError("recording changed while preparing accepted audio")
            data = output.getvalue()
            return data, hashlib.sha256(data).hexdigest()

    def review(self, run_id, take_id, request):
        repo = self.repository
        with repo.db.connect() as db:
            db.execute("BEGIN IMMEDIATE")
            take = repo.get_run_event(take_id)
            payload = take.payload
            if (take.run_id != run_id or take.event_type != "episode_audio_take_attached"
                    or payload.get("take_sha256") != request.expected_take_sha256
                    or digest({k: v for k, v in payload.items() if k != "take_sha256"}) != request.expected_take_sha256):
                raise ConflictError("listening review requires the exact saved recording take")
            current_takes = EpisodeAudioService(repo, self.data_root).recover(
                run_id, payload["sound_plan_id"], payload["expected_sound_plan_sha256"], _db=db)
            if not any(item.id == take_id for item in current_takes):
                raise ConflictError("recording choice changed; listen to the current take")
            events = repo.list_run_events(run_id)
            reviews = [event for event in events if event.event_type == "episode_audio_take_reviewed"
                       and event.payload.get("sound_plan_id") == payload["sound_plan_id"]
                       and event.payload.get("shot_index") == payload["shot_index"]]
            current = reviews[-1] if reviews else None
            if current and current.payload.get("review_sha256") != digest({k: v for k, v in current.payload.items() if k != "review_sha256"}):
                raise ConflictError("saved listening review integrity failed")
            request_sha = digest({"take_id": take_id, "request": request.model_dump()})
            if current and current.payload.get("request_sha256") == request_sha:
                return current
            if (current.id if current else None) != request.expected_review_id:
                raise ConflictError("listening decision changed; reload before confirming")
            if any("master" in event.event_type or event.event_type == "postproduction_materialized" for event in events):
                raise ConflictError("master work exists; reconcile before changing recording decisions")
            record = {"take_id": take_id, "take_sha256": request.expected_take_sha256,
                      "sound_plan_id": payload["sound_plan_id"], "sound_plan_sha256": payload["expected_sound_plan_sha256"],
                      "edit_review_id": payload["edit_review_id"], "edit_sha256": payload["edit_sha256"],
                      "shot_index": payload["shot_index"], "asset_id": payload["asset_id"],
                      "asset_sha256": payload["expected_asset_sha256"], "source_in_seconds": payload["source_in_seconds"],
                      "duration_seconds": payload["duration_seconds"], "decision": request.decision,
                      "reviewed_by": request.reviewed_by, "confirmation": request.confirmation,
                      "request_sha256": request_sha, "take_approved": request.decision == "accept",
                      "listening_evidence": "USER_ATTESTATION_NOT_PLAYBACK_TELEMETRY",
                      "speech_alignment_verified": False, "final_mix_approved": False,
                      "captions_approved": False, "master_accepted": False, "generation_performed": False}
            record["review_sha256"] = digest(record)
            identity = new_id("evt")
            sequence = db.execute("SELECT COALESCE(MAX(sequence),0)+1 FROM run_events WHERE run_id=?", (run_id,)).fetchone()[0]
            db.execute("INSERT INTO run_events VALUES (?,?,?,?,?,?,?,?,?)", (identity, run_id, sequence,
                       "episode_audio_take_reviewed", None, None, "Listening decision saved; alignment and final mix QA remain required.",
                       encode(record), utc_now()))
        return repo.get_run_event(identity)
